enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,374 @@
+console operator config.
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - service: my-console-plugin
+    namespace: openshift-operators
+    port: 8443
+    manifest: /manifest.json
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
+
+#### Alternative: ConsolePlugin CRD
+
+Rather than tying plugins to operators, we could have a CRD that allows you to
+declare a plugin.
+
+```yaml
+apiVersion: console.openshift.io/v1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
+  displayName: My Plugin
+  service:
+    name: my-console-plugin
+    namespace: my-operator-namespace
+    port: 8443
+    manifest: /manifest.json
+```
+
+Even if we have the CRD, we'd still want the cluster admin to enable the plugin
+before it becomes active.
+
+A disadvantage to this approach is that it's harder to limit plugins to Red
+Hat operators. It's also harder to trace a plugin back to an operator that
+owns it.
+
+### Loading Plugins
+
+The console backend will proxy requests to the plugin `Service`. This way the
+`Service` itself does not need to be exposed outside the cluster, and avoids
+any same-origin or certificate errors in the browser loading JavaScript from a
+separate domain. 
+
+On load, console will read the manifest for each plugin. The manifest is a JSON
+document that contains metadata about the plugin and the extensions.
+`extensions` mirrors the API used for static plugins, and the same extension
+points will be available.
+
+```json
+{
+  "$schema": "/path/to/schema/manifest.json",
+  "id": "my-plugin",
+  "version": "1.2.3",
+  "pluginAPI": "4.7.x",
+  "displayName": "My Plugin",
+  "extensions": [
+    {
+      "type": "console.flag",
+      "properties": {
+        "handler": {
+          "$codeRef": "example.testHandler"
+        }
+      }
+    },
+    {
+      "type": "console.flag/model",
+      "properties": {
+        "flag": "EXAMPLE",
+        "model": {
+          "group": "example.com",
+          "version": "v1",
+          "kind": "ExampleModel"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Plugin Manifest Properties
+
+| Property | Description |
+| ----- | -------- |
+| `name: string` | Used as a unique identifier. Each plugin must have a unique name. |
+| `version: string` | The version of the plugin. Version must be parsable by node-semver, eg. “1.2.3” |
+| `pluginAPI: string` | Semver range of compatible OpenShift console versions. |
+| `displayName?: string` | User friendly display name. |
+| `description?: string` | The description of the plugin. |
+| `extensions: Extension[]` | List of extensions contributed by the plugin. |
+| `dependencies?: { [pluginName: string]: string }` | A plugin which depends on other plugins must explicitly define their dependencies. |
+
+##### Extension<P = JSON>
```

Comment:
What is `<P = JSON>` about?  Possibly some web thing that anyone writing a plugin would recognize, but a link or text or something unpacking it for the rest of us would be nice :)